core/metacling/src/TCling.cxx

```diff
@@ -5891,11 +5867,85 @@ Int_t TCling::AutoParse(const char *cls)
    return nHheadersParsed > 0 ? 1 : 0;
 }
 
+static void* LazyFunctionCreatorAutoloadForModule(const std::string& mangled_name,
+      cling::Interpreter *fInterpreter) {
+   using namespace llvm::object;
+   using namespace llvm::sys::path;
+   using namespace llvm::sys::fs;
+
+   R__LOCKGUARD(gInterpreterMutex);
+
+   static bool sFirstRun = true;
+   // We don't want to do directory_iterator several times because it contains HDD access
+   static std::set<std::string> Libraries;
+
+   if (sFirstRun) {
+      // Store the information of path so that we don't have to iterate over the same path again and again.
+      std::unordered_set<std::string> AlreadyLookedPath;
+      const clang::Preprocessor &PP = fInterpreter->getCI()->getPreprocessor();
+      const HeaderSearchOptions &HSOpts = PP.getHeaderSearchInfo().getHeaderSearchOpts();
+      const auto ModulePaths(HSOpts.PrebuiltModulePaths);
+      cling::DynamicLibraryManager* dyLibManager = fInterpreter->getDynamicLibraryManager();
+
+      // Take path here eg. "/home/foo/module-release/lib/"
+      for (auto Path : ModulePaths) {
+         // Already searched?
+         if (AlreadyLookedPath.count(Path))
+            continue;
+
+         AlreadyLookedPath.insert(Path);
+         StringRef DirPath(Path);
+         if (!is_directory(DirPath))
+            continue;
+
+         std::error_code EC;
+         for (llvm::sys::fs::directory_iterator DirIt(DirPath, EC), DirEnd;
+               DirIt != DirEnd && !EC; DirIt.increment(EC)) {
+
+            std::string FileName(DirIt->path());
+            if (!llvm::sys::fs::is_directory(FileName) && extension(FileName) == ".so") {
+               // FIXME: TCling::IsLoaded is incredibly slow!
+               if (dyLibManager->isLibraryLoaded(FileName.c_str()))
+                  continue;
+               Libraries.insert(FileName);
+            }
+         }
+      }
+      sFirstRun = false;
+   }
+
+   // Iterate over files under this path. We want to get each ".so" files
+   for (std::string LibName : Libraries) {
+      auto SoFile = ObjectFile::createObjectFile(LibName);
+      if (SoFile) {
+         auto RealSoFile = SoFile.get().getBinary();
+         auto Symbols = RealSoFile->symbols();
+         for (auto S : Symbols) {
+            // DO NOT insert to table if symbol was weak or undefined
+            if (S.getFlags() == SymbolRef::SF_Weak || S.getFlags() == SymbolRef::SF_Undefined) continue;
+
+            if (S.getName().get() == mangled_name) {
```

Comment:
Please see how to handle llvm::Expected [here](http://llvm.org/docs/ProgrammersManual.html#recoverable-errors)